### PR TITLE
DROID-2738 Sync status | Status caching logic

### PR DESCRIPTION
**Pull request by konstantiniiv**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: data/src/main/java/com/anytypeio/anytype/data/auth/status/SyncAndP2PStatusEventsStore.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.anytypeio.anytype.data.auth.status
+
+import com.anytypeio.anytype.core_models.multiplayer.P2PStatusUpdate
+import com.anytypeio.anytype.core_models.multiplayer.SpaceSyncUpdate
+import kotlinx.coroutines.flow.Flow
+
+interface SyncAndP2PStatusEventsStore {
+    val p2pStatus: Flow<Map<String, P2PStatusUpdate>>
```
Comment (uburoiubu) [inline data/src/main/java/com/anytypeio/anytype/data/auth/status/SyncAndP2PStatusEventsStore.kt:8]: do you need a `Map` here or `Pair` ?

Comment (konstantiniiv) [reply data/src/main/java/com/anytypeio/anytype/data/auth/status/SyncAndP2PStatusEventsStore.kt:8]: Map

File: middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/AppEventChannel.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.anytypeio.anytype.middleware.interactor
+
+import anytype.Event
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+
+interface AppEventChannel {
```
Comment (konstantiniiv) [inline middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/AppEventChannel.kt:8]: event handler channel

File: middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/AppEventChannel.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    fun flow(): Flow<Event>
+    suspend fun emit(event: Event)
+}
+
+class AppEventChannelImpl(
+    replay: Int = 0,
```
Comment (konstantiniiv) [inline middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/AppEventChannel.kt:14]: remove from constructor

File: middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/SyncAndP2PChannelContainer.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private val _syncStatus = MutableStateFlow<MutableMap<String, SpaceSyncUpdate>>(mutableMapOf())
+
+    override val p2pStatus: Flow<Map<String, P2PStatusUpdate>> get() = _p2pStatus
+    override val syncStatus: Flow<Map<String, SpaceSyncUpdate>> get() = _syncStatus
+
+    private val jobs = mutableListOf<Job>()
+
+    override fun start() {
+        Timber.i("SyncAndP2PStatusEventsStoreImpl start")
+        jobs.forEach { it.cancel() }
```
Comment (konstantiniiv) [inline middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/SyncAndP2PChannelContainer.kt:34]: ext func

File: middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/SyncAndP2PChannelContainer.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        jobs.clear()
+        jobs += scope.launch(Dispatchers.IO) {
```
Comment (konstantiniiv) [inline middleware/src/main/java/com/anytypeio/anytype/middleware/interactor/SyncAndP2PChannelContainer.kt:36]: move dispatcher to constructor

**Review by uburoiubu — APPROVED** [review]

**Pull request merged**